packages/node/src/channels.ts

```diff
@@ -0,0 +1,387 @@
+import { legacy } from "@counterfactual/cf.js";
+import {
+  Address,
+  AppInstanceInfo,
+  AssetType,
+  Node
+} from "@counterfactual/common-types";
+import { ethers } from "ethers";
+
+import { IStoreService } from "./service-interfaces";
+import { zeroBalance } from "./utils";
+
+import Nonce = legacy.utils.Nonce;
+import FreeBalance = legacy.utils.FreeBalance;
+
+/**
+ * The schema of a channel is below.
+ * The following Channels class encapsulates the state and persistence of all
+ * the channels in a Node.
+ */
+
+/**
+ * The fully expanded schema for a channel:
+ * multisigAddress: {
+ *  multisigAddress: Address,
+ *  multisigOwners: Address[],
+ *  appsNonce: Nonce,
+ *  appInstances: Map<AppInstanceID,
+ *    appInstance: {
+ *      id: string,
+ *      appId: Address,
+ *      abiEncodings: {
+ *        stateEncoding: string,
+ *        actionEncoding: string
+ *      },
+ *      appState: any,
+ *      localNonce: Nonce,
+ *      dependencyNonce: Nonce,
+ *      timeout: BigNumber,
+ *      asset: {
+ *        assetType: AssetType,
+ *        token?: Address
+ *      },
+ *      deposits: Map<Address, BigNumber>
+ *    }
+ *  },
+ *  proposedAppInstances: same schema as appInstances,
+ *  freeBalances: Map<AssetType,
+ *    freeBalance: {
+ *      alice: Address,
+ *      aliceBalance: BigNumber,
+ *      bob: Address,
+ *      bobBalance: BigNumber,
+ *      uniqueId: number,
+ *      localNonce: number,
+ *      timeout: number,
+ *      dependencyNonce: {
+ *        isSet: boolean,
+ *        salt: string,
+ *        nonceValue: number
+ *      }
+ *    }
+ *  }
+ * }
+ */
+
+/**
+ * This class is only a type implementation of a channel schema for the
+ * purposes of updating and retrieving a channel's state from the store.
+ *
+ * An instance is by itself stateless and effectively reflects the state of an
+ * according channel in the store.
+ */
+class Channel {
+  constructor(
+    readonly multisigAddress: Address,
+    readonly multisigOwners: Address[],
+    readonly appsNonce: Nonce = new Nonce(true, 0, 0),
```

Comment:
we should not have a default value here and specify it explicitly at constructor call time instead

Reply:
How come?

This only exists to load a pre-existing `appsNonce` upon querying the store for that channel. If no pre-existing channel exists then this default seems natural?

Reply:
is it a property that you must specify a default (or a default-constructable type, or something) in the store schema?

Reply:
Not right now.

Though what's the use case for not having a default here when a new channel is being created? What's the benefit of having the caller supply it when they're creating a _new_ channel?

Reply:
The root nonce timeout for a new channel shouldn't be 0, for one